// states-nepal/lib/esm/src/entities/category.js
import { fetcher } from '../fetchers';
/**
 * Class Category
 */
export class Category {
    /**
     * Category constructor.
     * @param string lang
     * @throws exception
     */
    constructor(lang = 'en') {
        try {
            this.lang = lang;
            this.categories = fetcher('categories', this.lang);
        }
        catch (err) {
            throw new Error(`Categories of given language doesn't exists.`);
        }
    }
    /**
     * Get list of all categories
     *
     * @returns categories array
     */
    allCategories() {
        return this.categories;
    }
    /**
     * Find category deatils by unique id
     *
     * @param id
     * @returns category | null
     */
    find(id) {
        const category = this.categories.find(el => el.id === id);
        return category ? category : null;
    }
    /**
     * Find category details by short code
     *
     * @param short_code
     * @return category | null
     */
    findByShortCode(short_code) {
        const category = this.categories.find(el => el.short_code === short_code);
        return category ? category : null;
    }
    /**
     * Search Categories
     *
     * @param key
     * @param value
     * @return array of categories that match with given key
     */
    search(key, value) {
        return this.categories.filter(el => (el[key] ? el[key] === value : null));
    }
}
