// states-nepal/lib/esm/src/entities/district.js
import { fetcher } from '../fetchers';
import { numericEnglish } from '../utils';
import { Municipality } from './municipality';
/**
 * Class District
 */
export class District {
    /**
     * District constructor.
     * @param string lang
     * @throws exception
     */
    constructor(lang = 'en') {
        try {
            this.lang = lang;
            this.districts = fetcher('districts', this.lang);
        }
        catch (err) {
            throw new Error(`Districts of given language doesn't exists. `);
        }
    }
    /**
     * Get list of all districts
     *
     * @returns districts array
     */
    allDistricts() {
        return this.districts;
    }
    /**
     * Find district deatils by unique id
     *
     * @param id
     * @returns district | null
     */
    find(id) {
        const district = this.districts.find(el => el.id === id);
        return district ? district : null;
    }
    /**
     * Get districts by province id
     *
     * @param provinceId
     * @return array|null
     */
    getDistrictsByProvince(provinceId) {
        return this.districts.filter(el => el.province_id === provinceId);
    }
    /**
     * Get district with largest area
     *
     * @return district
     */
    largest() {
        const areas = this.districts.map(el => el.area_sq_km);
        let numericArea = [];
        if (this.lang === 'np') {
            numericArea = areas.map(el => numericEnglish(el)).map(i => Number(i));
        }
        return this.districts[numericArea.indexOf(Math.max(...numericArea))];
    }
    /**
     * Get district with smallest area
     *
     * @return district
     */
    smallest() {
        const areas = this.districts.map(el => el.area_sq_km);
        let numericArea = [];
        if (this.lang === 'np') {
            numericArea = areas.map(el => numericEnglish(el)).map(i => Number(i));
        }
        return this.districts[numericArea.indexOf(Math.min(...numericArea))];
    }
    /**
     * Get district with municipalities
     *
     * @return array of districts with municipalities
     */
    getDistrictsWithMunicipalities() {
        const municipality = new Municipality(this.lang);
        const districts = this.allDistricts();
        const districtsWithMunicipalities = districts.map(districtItem => {
            var _a;
            return (Object.assign(Object.assign({}, districtItem), { municipalities: (_a = municipality
                    .getMunicipalitiesByDistrict(districtItem.id)) === null || _a === void 0 ? void 0 : _a.map(municipalityItem => (Object.assign(Object.assign({}, municipalityItem), { wards: municipality.wards(municipalityItem.id) }))) }));
        });
        return districtsWithMunicipalities;
    }
    /**
     * Search Districts
     *
     * @param key
     * @param value
     * @return array of districts that match with given key
     */
    search(key, value) {
        return this.districts.filter(el => (el[key] ? el[key] === value : null));
    }
}
