// states-nepal/lib/esm/src/entities/province.js
import { fetcher } from '../fetchers';
import { numericEnglish } from '../utils';
import { District } from './district';
import { Municipality } from './municipality';
/**
 * Class Province
 */
export class Province {
    /**
     * Province constructor.
     * @param string lang
     * @throws exception
     */
    constructor(lang = 'en') {
        try {
            this.lang = lang;
            this.provinces = fetcher('provinces', this.lang);
        }
        catch (err) {
            throw new Error(`Province of given language doesn't exists.`);
        }
    }
    /**
     * Get list of all provinces
     *
     * @returns provinces array
     */
    allProvinces() {
        return this.provinces;
    }
    /**
     * Find province deatils by unique id
     *
     * @param id
     * @returns province | null
     */
    find(id) {
        const province = this.provinces.find(el => el.id === id);
        return province ? province : null;
    }
    /**
     * Get province with largest area
     *
     * @return province
     */
    largest() {
        const areas = this.provinces.map(el => el.area_sq_km); // areas are returned in string value
        let numericArea = [];
        if (this.lang === 'np') {
            numericArea = areas.map(el => numericEnglish(el)).map(i => Number(i));
        }
        return this.provinces[numericArea.indexOf(Math.max(...numericArea))];
    }
    /**
     * Get province with smallest area
     *
     * @return province
     */
    smallest() {
        const areas = this.provinces.map(el => el.area_sq_km);
        let numericArea = [];
        if (this.lang === 'np') {
            numericArea = areas.map(el => numericEnglish(el)).map(i => Number(i));
        }
        return this.provinces[numericArea.indexOf(Math.min(...numericArea))];
    }
    /**
     * Get provinces with districts
     *
     * @return array of provinces with districts
     */
    getProvincesWithDistricts() {
        const district = new District(this.lang);
        const provinces = this.provinces;
        return provinces.map(provinceItem => (Object.assign(Object.assign({}, provinceItem), { districts: district.getDistrictsByProvince(provinceItem.id) })));
    }
    /**
     * Get provinces with districts with its municipalities
     *
     * @return array of provinces
     */
    getProvincesWithDistrictsWithMunicipalities() {
        const district = new District(this.lang);
        const municipality = new Municipality(this.lang);
        const provinces = this.provinces;
        return provinces.map(provinceItem => (Object.assign(Object.assign({}, provinceItem), { districts: district
                .getDistrictsByProvince(provinceItem.id)
                .map(districtItem => {
                var _a;
                return (Object.assign(Object.assign({}, districtItem), { municipalities: (_a = municipality
                        .getMunicipalitiesByDistrict(districtItem.id)) === null || _a === void 0 ? void 0 : _a.map(municipalityItem => (Object.assign(Object.assign({}, municipalityItem), { wards: municipality.wards(municipalityItem.id) }))) }));
            }) })));
    }
    /**
     * Search provinces
     *
     * @param key
     * @param value
     * @return array of provinces that match with given key
     */
    search(key, value) {
        return this.provinces.filter(el => (el[key] ? el[key] === value : null));
    }
}
