// states-nepal/lib/cjs/src/entities/municipality.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Municipality = void 0;
const fetchers_1 = require("../fetchers");
const utils_1 = require("../utils");
/**
 * Class Municipality
 */
class Municipality {
    /**
     * Category constructor.
     * @param string lang
     * @throws exception
     */
    constructor(lang = 'en') {
        try {
            this.lang = lang;
            this.municipalities = (0, fetchers_1.fetcher)('municipalities', this.lang);
        }
        catch (err) {
            throw new Error(`Municipalities of given language doesn't exists.`);
        }
    }
    /**
     * Get list of all municipalities
     *
     * @returns municipalities array
     */
    allMunicipalities() {
        return this.municipalities;
    }
    /**
     * Get municipalities by district id
     *
     * @param districtId
     * @returns array of municipalities | null
     */
    getMunicipalitiesByDistrict(districtId) {
        const municipalities = this.municipalities.filter(el => el.district_id === districtId);
        return municipalities ? municipalities : null;
    }
    /**
     * Get municipalities by category Id
     *
     * @param categoryId
     * @returns array of municipalities | null
     */
    getMunicipalityByCategory(categoryId) {
        const municipalities = this.municipalities.filter(el => el.category_id === categoryId);
        return municipalities ? municipalities : null;
    }
    /**
     * Find municipalities by id
     *
     * @param id
     * @returns municipality | null
     */
    find(id) {
        const municipality = this.municipalities.find(el => el.id === id);
        return municipality ? municipality : null;
    }
    /**
     * Get municipality with largest area
     *
     * @return municipality
     */
    largest() {
        const areas = this.municipalities.map(el => el.area_sq_km); // areas are returned in string value
        let numericArea = [];
        if (this.lang === 'np') {
            numericArea = areas.map(el => (0, utils_1.numericEnglish)(el)).map(i => Number(i));
        }
        return this.municipalities[numericArea.indexOf(Math.max(...numericArea))];
    }
    /**
     * Get municipality with smallest area
     *
     * @return municipality
     */
    smallest() {
        const areas = this.municipalities.map(el => el.area_sq_km);
        let numericArea = [];
        if (this.lang === 'np') {
            numericArea = areas.map(el => (0, utils_1.numericEnglish)(el)).map(i => Number(i));
        }
        return this.municipalities[numericArea.indexOf(Math.min(...numericArea))];
    }
    /**
     * Get wards of municipality
     *
     * @param id
     * @return array of wards
     */
    wards(id) {
        const municipality = this.find(id);
        if (this.lang === 'np' && municipality) {
            const totalWards = parseInt((0, utils_1.numericEnglish)(municipality.wards));
            const wards = (0, utils_1.range)(1, totalWards);
            return wards.map(i => String(i)).map(el => (0, utils_1.numericNepali)(el));
        }
        return (municipality && (0, utils_1.range)(1, parseInt(municipality.wards)).map(i => String(i)));
    }
    /**
     * Search Municipalities
     *
     * @param key
     * @param value
     * @return array of municipalities that match with given key
     */
    search(key, value) {
        return this.municipalities.filter(el => el[key] ? el[key] === value : null);
    }
}
exports.Municipality = Municipality;
